### PR TITLE
[c10d] Fixed MPI build with higher version of GCC

### LINKED ISSUE #14078: Build failing at torch/lib/c10d/ProcessGroupMPI.cpp
## 🐛 Bug

PyTorch fails to finish building, with a possible bug (see below).

```bash 
Scanning dependencies of target caffe2_observers
[ 92%] Building CXX object caffe2/torch/lib/THD/CMakeFiles/THD.dir/base/data_channels/DataChannelTCP.cpp.o
[ 93%] Building CXX object modules/observers/CMakeFiles/caffe2_observers.dir/net_observer_reporter_print.cc.o
[91m/opt/pytorch/torch/lib/c10d/ProcessGroupMPI.cpp: In destructor ‘virtual c10d::ProcessGroupMPI::AsyncWork::~AsyncWork()’:
/opt/pytorch/torch/lib/c10d/ProcessGroupMPI.cpp:154:71: error: throw will always call terminate() [-Werror=terminate]
         "Attempted destruction of AsyncWork before work has completed");
                                                                       ^
/opt/pytorch/torch/lib/c10d/ProcessGroupMPI.cpp:154:71: note: in C++11 destructors default to noexcept
[0m[ 93%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/function.cpp.o
[ 93%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/functions/accumulate_grad.cpp.o
[ 93%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/ps_roi_pool_op.cc.o
[ 93%] Building CXX object modules/observers/CMakeFiles/caffe2_observers.dir/observer_config.cc.o
[91mcc1plus: all warnings being treated as errors
[0m[91mmake[2]: *** [caffe2/torch/lib/c10d/CMakeFiles/c10d.dir/ProcessGroupMPI.cpp.o] Error 1
[0m[91mmake[2]: *** Waiting for unfinished jobs....
[0mcaffe2/torch/lib/c10d/CMakeFiles/c10d.dir/build.make:230: recipe for target 'caffe2/torch/lib/c10d/CMakeFiles/c10d.dir/ProcessGroupMPI.cpp.o' failed
[ 93%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/functions/basic_ops.cpp.o
[ 93%] Building CXX object caffe2/torch/lib/THD/CMakeFiles/THD.dir/base/init_methods/InitMethod.cpp.o
[ 93%] Building CXX object modules/observers/CMakeFiles/caffe2_observers.dir/perf_observer.cc.o
[ 93%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/roi_pool_f_op.cc.o
[ 93%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/sample_as_op.cc.o
[ 93%] Building CXX object caffe2/torch/lib/THD/CMakeFiles/THD.dir/base/init_methods/InitMethodEnv.cpp.o
[ 93%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/functions/comm.cpp.o
[ 93%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/select_smooth_l1_loss_op.cc.o
[ 93%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/functions/tensor.cpp.o
CMakeFiles/Makefile2:7496: recipe for target 'caffe2/torch/lib/c10d/CMakeFiles/c10d.dir/all' failed
[91mmake[1]: *** [caffe2/torch/lib/c10d/CMakeFiles/c10d.dir/all] Error 2
[0m[91mmake[1]: *** Waiting for unfinished jobs....
...
...
...
[ 99%] Linking CXX shared library ../../lib/libtorch.so
[ 99%] Built target torch
Makefile:138: recipe for target 'all' failed
[91mmake: *** [all] Error 2
[0msetup.py::build_deps::run()
Failed to run 'bash ../tools/build_pytorch_libs.sh --use-fbgemm --use-nnpack --use-mkldnn --use-qnnpack caffe2'
```

## To Reproduce
```bash 
 cd /opt && git clone --recursive https://github.com/pytorch/pytorch \
    && cd pytorch && git submodule update --init && \
    cd /opt/pytorch/third_party/ideep/mkl-dnn && \
    git pull https://github.com/intel/mkl-dnn.git --no-commit  --rebase && \
    cd /opt/pytorch && \
    sed -i 's/"Use MKLDNN" OFF/"Use MKLDNN" ON /g' CMakeLists.txt && \
    sed -i 's/"Use DISTRIBUTED" OFF/"Use DISTRIBUTED" ON /g' CMakeLists.txt && \
    sed -i 's/for parallel code" OFF/for parallel code" ON /g' CMakeLists.txt && \
    PYTHON_EXECUTABLE=/opt/conda/bin/python \
    PYTHON_LIBRARY=/opt/conda/lib/libpython3.6m.so \
    PYTHON_INCLUDE_DIR=/opt/conda/include/python3.6m \
    FULL_CAFFE2=1 \
    USE_OPENMP=1 \
    USE_MKL=1 \
    USE_MKLDNN=1 \
    USE_MKLML=1 \
    USE_SYSTEM_EIGEN_INSTALL=1 \
    USE_ZMQ=1 \
    USE_DISTRIBUTED=1 \
    MKLDNN_LIBRARY=/usr/local/lib \
    MKLDNN_INCLUDE_DIR=/usr/local/include \
    MKLDNN_LIB_DIR=/usr/local/lib \
    python setup.py install && \
    cd /opt && rm -rf /opt/pytorch && \
    cd /usr/lib && sudo ldconfig
```

## Environment

 - PyTorch Version (e.g., 1.0): Master branch (1.0) 
 - OS (e.g., Linux): Debian:Stretch
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): see above
 - Python version: 3.6
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

### PR DESCRIPTION
**Pull request by teng-li**
This appears as I enabled -Werror in c10d build. Good to catch this and fix it.

Should fix https://github.com/pytorch/pytorch/issues/14078 and https://github.com/pytorch/pytorch/issues/13962

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupMPI.cpp [hunk 1]
```diff
@@ -150,8 +150,11 @@ ProcessGroupMPI::AsyncWork::AsyncWork(
 
 ProcessGroupMPI::AsyncWork::~AsyncWork() {
   if (request_ != MPI_REQUEST_NULL) {
-    throw std::runtime_error(
-        "Attempted destruction of AsyncWork before work has completed");
+    std::cerr
+        << "Attempted destruction of AsyncWork before work has completed, "
+        << "terminating the program."
+        << std::endl;
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupMPI.cpp]: Please add a std::flush as well to be sure it is written to the underlying STDERR

**Review by pietern — APPROVED** [review]

File: torch/lib/c10d/ProcessGroupMPI.cpp (hunk 1 above)
Comment (teng-li) [reply torch/lib/c10d/ProcessGroupMPI.cpp]: std::endl will call os.flush(). Adding one more CI change before landing

**Comment by ezyang** [comment]
@teng-li -Werror should never be enabled for user builds. We can enable it in CI so that we can catch warnings for ourselves, but we never know what kind of compiler someone is going to use (and what errors they toggle by default), and setting -Werror in this case is just asking for people to be broken for no good reason.

**Comment by soumith** [comment]
I agree with Edward. We should just set it in CI, not uniformly.

**Comment by teng-li** [comment]
https://github.com/pytorch/pytorch/pull/14155 should disable it

**Pull request closed without merging**